File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    let max = -Infinity
    function helper(node) {
        
        if (!node) {
            return [Infinity,-Infinity]
        }
        let left = helper(node.left)
        let right = helper(node.right)
        let d1 = Math.abs(node.val - left[0])
        let d2 = Math.abs(node.val - left[1])
        let d3 = Math.abs(node.val - right[0])
        let d4 = Math.abs(node.val - right[1])
        if (d1 == Infinity) d1 = -Infinity
        if (d2 == Infinity) d2 = -Infinity
        if (d3 == Infinity) d3 = -Infinity
        if (d4 == Infinity) d4 = -Infinity
        max = Math.max(d1,d2,d3,d4,max)
        return [Math.min(left[0], node.val, right[0]), Math.max(right[1], node.val, left[1])]
    }
    helper(root)
    return max
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.diff = 0

    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            if not root:
                return [sys.maxsize, -sys.maxsize]

            if root.left == None and root.right == None:
                return [root.val, root.val]
            
            left = dfs(root.left)
            right = dfs(root.right)

            minVal = min(left[0], right[0])
            maxVal = max(left[1], right[1])
                
            self.diff = max(self.diff, max(abs(minVal - root.val), abs(maxVal - root.val)))

            minVal = min(minVal, root.val)
            maxVal = max(maxVal, root.val)

            return [minVal, maxVal]
        
        dfs(root)

        return self.diff 
        